### ISSUE TITLE
Problems with capitalized file extensions

### ISSUE THREAD

**Issue by scval** [git-lfs/git-lfs#2858]
On Windows, I migrated a non-lfs git repository to git lfs using git lfs 2.3.4:
`git lfs migrate import --include="*.exe" --include-ref=refs/heads/master`
I expected that the migration is carried out for all possibilities of capitalization: *.exe, *.Exe, *.EXE, i.e. case-insensitive pattern matching. This is what I (naively) expected at least on a (partly) case-insensitive file system like NTFS on Windows.

What really happens after executing the command stated above is that only *.exe files are migrated to git lfs, *.EXE files are not migrated.

On the other hand, when setting up a new git repository using git lfs and setting .gitattributes to
`*.exe filter=lfs diff=lfs merge=lfs -text`
then committing both a file ending on  *.exe and on *.EXE leads to the equivalent result with respect to git lfs, so each file is committed as a pointer and stored using git lfs.

Is this the expected behavior? I did not check on Linux and/or a case-sensitive filesystem so far.

Thanks for any input.

**Comment by ttaylorr** [comment]
Hi @scval, thanks for opening this issue. I did not previously know about this issue, but the behavior is in line with my best knowledge of the `filepathfilter` package (which is used to match patterns like `*.exe` with files on disk).

I am in the process of changing this package significantly, and respecting filesystem case-sensitivity has been addressed. I am hoping to include these changes in the next release of Git LFS, so I am going to close this issue for now. Thanks!

**Comment by scval** [comment]
Hi @ttaylorr,
thank you for your answer! Sounds great that you are going to address the issue in the next release!

One question/comment: I do not understand what the desired behavior is. Currently, the behavior seems to be different when migrating to lfs (extensions are handled case-sensitive) and when doing new commits with lfs (extensions are handled case-insensitive, what I find more appropriate for case-intensitive Windows/NTFS).
Any clarification would be highly appreciated. Thank you very much!

Second question: Shall .gitattributes explicitly name all possibilities of capitalization (e.g. *.exe, *.EXE, *.Exe, ...) for each tracked lfs file type? Unfortunately, Windows users usually do not care on file name capitalization.

Why this issue is so important for us is that git lfs seems not handle non-migrated lfs files very well. In our case, a file with capitalized extension (EXE) was not migrated. In .gitattributes, only *.exe is named. Then, git lfs leads to spurious modifications and warnings “Encountered 1 file(s) that should have been pointers, but weren't ”, that cannot be reset using
git reset —hard
or
git clean -fdx
but only using git lfs uninstall followed by git status and git lfs install.

Thank you again for any support! And please let me know if I can provide you with some test data.

**Comment by ttaylorr** [comment]
> One question/comment: I do not understand what the desired behavior is. Currently, the behavior seems to be different when migrating to lfs (extensions are handled case-sensitive) and when doing new commits with lfs (extensions are handled case-insensitive, what I find more appropriate for case-intensitive Windows/NTFS).

Case sensitivity will respect the system defaults in all places now, not just in the migrator.



> Second question: Shall .gitattributes explicitly name all possibilities of capitalization (e.g. *.exe, *.EXE, *.Exe, ...) for each tracked lfs file type? Unfortunately, Windows users usually do not care on file name capitalization.

I believe not, but @larsxschneider would have to clarify this for me.

**Comment by larsxschneider** [comment]
@scval The `.gitattributes` filter uses native Git mechanics and is indeed case insensitive on Windows/NTFS. However, your repository would run into trouble on a case sensitive filesystem (e.g. Linux) as Git LFS would only process the `.exe` files and leave the `.EXE` files alone (see my [talk](https://www.youtube.com/watch?v=YQzNfb4IwEY) on the topic for details). Therefore, I recommend to always use case insensitve patterns like this:
```
*.[eE][xX][eE]
```
(that is a so called "glob" pattern. You can name the different casing explicitly too)

The `git lfs migrate`, however, accesses the Git database directly and therefore all file checks are case sensitive (as Git is internally case sensitive).

I hope this makes some sense 😄

**Comment by ttaylorr** [comment]
> I hope this makes some sense 😄

Thanks for the clarification!

**Comment by scval** [comment]
Hello @larsxschneider,

thank you for your reply and for posting the video of the talk! And sorry for the late answer.

Indeed, the explanations in your talk helped, and we use glob patterns now in our gitattributes as you proposed! Now it seems to work as expected.

However, I still had many (basically unrelated) problems with git lfs when doing all the conversion/migration work for fixing the errors stated in this issue (git filter-branch, cherry-pick, etc.). It seems to me that it is a problem for git lfs if a file is not checked in using git lfs (no lfs pointer, but checked into the regular git repository) and the file is matched by a pattern in .gitattributes for being tracked by lfs.
The equivalent applies for the case that a file is not mentioned in .gitattributes (no lfs tracking), but checked in using git lfs (hence, using a lfs pointer and in the lfs space). Unfortunately, these inappropriate states occurred for me.

Then, at least in my case, git lfs and git (`git status`) showed (tracked, unstaged) modified files, that cannot be reverted using `git reset --hard` or `git clean -fdx`. One workaround that helped me is
`git lfs uninstall && git status && git lfs install`
to get back to a clean state (no modified files anymore). Possibly, administrator permissions are required once for `git lfs uninstall`, what can be checked using `git lfs env` (lfs filters must be empty after uninstall).

Is this behavior known? It would be great if git lfs could be less picky here and be more robust with respect to such defective states.

Thanks for all the support, again!

**Comment by ttaylorr** [comment]
> Is this behavior known? It would be great if git lfs could be less picky here and be more robust with respect to such defective states.

It's a known behavior. The issue is that LFS has very little control over preventing these types of situations, because they often occur when Git LFS isn't installed, so we have no way to prevent them.

One thing that I think would be interesting to address in a subsequent release, I agree, is handling these cases better when they occur. For instance, LFS could try and see if a large file matched a pointer, and in that case, print out the command to migrate it correctly.

What do you think of that?

**Comment by scval** [comment]
Hello @ttaylorr,

thank you for your reply. I think your proposal is fine.

However, we also had problems when git lfs is installed all the time for one user/machine, while another user commits binaries without lfs being installed (be it due to an old version of git without lfs, or because lfs is not (correctly) installed).

Please consider the following situation, that is very realistic for our setting:
We have a large repository on a Bitbucket server. LFS is enabled for the repository and used for several binaries (e.g. *.[jJ][pP][gG]). Now one user commits and pushes a binary (that is indicated to be tracked by lfs according to .gitattributes) without having lfs installed on the machine or without having it properly configured/installed for the repository. This can happen for a large user base that push to the repository, and I do not know how we cannot avoid it centrally (e.g. on the server).

Now there is a binary in the git instead of the lfs link expected by git lfs.

This is a situation that seems not to be handled very well by git lfs. If we pull the changes from another repository with git lfs installed, we get a warning regarding the checked in binary (OK) and then spurious modified files (what I do not like). The same problem occurs when cloning this repository. First, just the expected warning is shown when cloning:
`$ git clone https://bitbucket_server_url/git_lfs_test.git git_lfs_test_defect_repo
Cloning into 'git_lfs_test_defect_repo'...
remote: Counting objects: 11, done.
remote: Compressing objects: 100% (9/9), done.
remote: Total 11 (delta 3), reused 0 (delta 0)
Unpacking objects: 100% (11/11), done.
Downloading test.jpg (5 B)
Checking out files: 100% (7/7), done.
Encountered 1 file(s) that should have been pointers, but weren't:
        test_new.jpg
`
But if we check the status, it shows
`$ git status
On branch master
Your branch is up-to-date with 'origin/master'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

        modified:   test_new.jpg

no changes added to commit (use "git add" and/or "git commit -a")`

Checking out the file again or resetting the changes does not help:
`$ git checkout -- test_new.jpg
Encountered 1 file(s) that should have been pointers, but weren't:
        test_new.jpg`
The status remains still modified. How can we fix this state?

`git lfs uninstall && git status && git lfs install`
does help, but this is only a workaround.

Thank you!

**Comment by larsxschneider** [comment]
@scval Try this: https://github.com/git-lfs/git-lfs/issues/2845#issuecomment-361523599

I am thinking about a command such as `git lfs repair $filepath` which would essentially do:
```
git rm --cached $filepath
git add --force $filepath
git commit -m "Move files properly to GitLFS"
```
Would that be an acceptable solution for you?

**Comment by ttaylorr** [comment]
> Would that be an acceptable solution for you?

I'm not sure that I'm in favor of adding this to Git LFS necessarily. My reasoning is that while the `rm --cached && add --force` combination will absolute reinstate the file into Git LFS, previous versions of history remain broken (so checking them out will still cause problems).

What do you think about adding this into the `git lfs migrate import` command? Perhaps something like a special mode: `git lfs migrate import --no-rewrite --include="path/to/file"`? In this mode, we could perform the above operation by adding a _new_ commit to Git LFS instead of rewriting the old one.

What do you think?

**Comment by scval** [comment]
Thank you again for your help!

Actually, I would prefer if git lfs could handle these somewhat inconsistent states (i.e. no lfs link, but binary file itself in git, although indicated in `.gitattribtues` for being tracked by git lfs) without doing a commit or a migration.

I understand that git lfs shows the message ``Encountered 1 file(s) that should have been pointers, but weren't: [...]``. This makes sense and is understandable from a user's perspective.

However, additionally, git shows a modified file in this case, but the binary is in sync with the corresponding git repository. The only inconsistency is that it is not a link. From a user's perspective, the modified state seems misleading to me. Furthermore, the fix mentioned by git status `(use "git checkout -- <file>..." to discard changes in working directory)` does not change the modified state, the file remains in modified state. So normal users cannot recover from this situation, and because of modified files, they cannot checkout/reset files, checkout/switch branches, merge/rebase, or pull in most cases.

If the unchanged binary is committed, it is no change in content, but only a repository change (migration to lfs). Furthermore, the inconsistent state remains in the history and might be checked out. This could confuse users. From this perspective, I would prefer migration, but this means rewriting history. This again might confuse users. Therefore I prefer a warning regarding the inconsistent state, but git lfs should remain operable (e.g. checkout, reset, switch, ...)

Can you help me here? Or is this not possible within the (current) architecture or implementation of git/git lfs?

Thank you very much!

**Comment by vizanto** [comment]
I just ran: `git lfs migrate import --include '*.[zZ][iI][pP]' --everything`
- The `.gitattributes` file contains the pattern
- No files were actually migrated. `git lfs ls-files` shows no output.

Then I ran: `git lfs migrate import --include '*.zip' --everything`
- `git lfs ls-files` shows zip files.

My guess is the migrate tool doesn't use the same glob/regex pattern as git?

**Comment by ttaylorr** [comment]
> I just ran: `git lfs migrate import --include '*.[zZ][iI][pP]' --everything`
> 
> * The `.gitattributes` file contains the pattern
> * No files were actually migrated. `git lfs ls-files` shows no output.

Hi @vizanto -- I just released a new version of Git LFS which supports the character class-style matching (the `[]`'s above). In previous versions of Git LFS, this was not possible. Would you mind upgrading Git LFS to the latest version and giving it another shot?

Everything should work as expected, which is why I'm closing this issue now, but if you run into any issues, please don't hesitate to ping me here and I'd be more than happy to get this issue back open.

**Comment by scval** [comment]
Hi @ttaylorr,

thank you for implementing support for using glob patterns/character class-style matching when doing the migration with git lfs! That seems to be a perfect solution for the original issue described by me above in my first comment.

However, the (somewhat unrelated but subsequent) second issue described in the last two comments of myself from above is still very critical for us (see above, https://github.com/git-lfs/git-lfs/issues/2858#issuecomment-369910325 and https://github.com/git-lfs/git-lfs/issues/2858#issuecomment-369186794).

This issue means that a single user without a proper installation of git lfs is able to crash the repository. For our large user base, this might occur from time to time, and leads to problems for all the other users (spurious modified files, unusable repository, like described above). Fixes have to be rolled out to all users, which I think is a large effort.

Since there is no possibility to check the git and git lfs installation on the machines of the users, I do not know of a possiblity how to avoid from these wrong commits of binaries itsselves (instead of lfs links). 

As a first workaround, we have enabled pull requests on the development branch on the Bitbucket server now because of this issue. However, I am not sure whether the reviewers can and do see these wrong (no lfs) commits in all cases. To me, this seems to be an unstable solution.

This issue might be related to https://github.com/git-lfs/git-lfs/issues/2874, but I did not see the same error message.

Is there a possibility how git lfs and its git integration can be improved here? I have the feeling that the criticality of this bug for us was not clear enough so far. It might render git lfs unusable for us.

Any further help would be highly appreciated!

Shall I open a new issue on this? What do you think?

Thank you very much.

**Comment by ttaylorr** [comment]
> That seems to be a perfect solution for the original issue described by me above in my first comment.

Wonderful; I'm glad to hear that it is working for you. 😄 

> Is there a possibility how git lfs and its git integration can be improved here? I have the feeling that the criticality of this bug for us was not clear enough so far. It might render git lfs unusable for us.

I agree that this is a problem, and you are correct in noting that it is difficult for the Git LFS client to catch this type of error before pushing changes to a remote. I think the most sensible thing that providers can do is reject pushes that contain Git blobs where that path should be tracked under Git LFS.

This is certainly difficult to do in practice, but I think is the only way one can be absolutely sure that repositories have not gotten into a bad state.

That said, the `git lfs migrate` tool is designed to help resolve this situation with relative ease (or, as much as is possible in this situation). 

> Shall I open a new issue on this? What do you think?

If you would like to do so, that would be OK.

**Comment by scval** [comment]
Can it be avoided that git with installed git lfs shows modified files for git blobs that should be tracked under git lfs? 
(see https://github.com/git-lfs/git-lfs/issues/2858#issuecomment-369186794)
From the warning `Encountered 1 file(s) that should have been pointers, but weren't: [...]`, it is clear that git lfs knows about the problem. But these modified files cannot be cleared. Only `git lfs uninstall && git status && git lfs install` helps. Then only the warning appears, no modified files anymore. This is the behavior that I would expect. Is there no way to automatically handle this?

Even worse, the workaround git lfs uninstall requires administrative access if git lfs was installed system-widely (like in our case).

Thank you!

**Comment by ttaylorr** [comment]
Hi @scval, thanks for waiting and sorry for the delay! Here are some answers to your questions:

> Can it be avoided that git with installed git lfs shows modified files for git blobs that should be tracked under git lfs?

It is possible for Git LFS to suppress this warning, but I am not sure that it should be done. In the case that you have described above, history _is_ broken, and needs either a (1) rewrite, or (2) a new commit fixing the working copy (as Lars proposes above).

Perhaps we could suppress this error iff the Git blob SHA matches the LFS pointer SHA. @larsxschneider I am not sure how I feel about this. Do you have any ideas?

> Even worse, the workaround git lfs uninstall requires administrative access if git lfs was installed system-widely (like in our case).

This is intended, since a user needs permission to elevate in order to delete a file owned by someone else.

**Comment by larsxschneider** [comment]
> Perhaps we could suppress this error iff the Git blob SHA matches the LFS pointer SHA. 
Unfortunately, I don't think this is an option as Git uses SHA-1 and Git LFS SHA-256.

I agree @ttaylorr that the history is broken in that case and needs to be fixed. I think we should not cover up these kind of errors.

**Comment by mojca** [comment]
I'm sorry for writing to an ancient thread, but I'm perplexed.

You seem to agree that history of a repository is broken and that the repository needs to be fixed. But doesn't that mean that there is a bug in `git lfs` if `git lfs migrate` itself generates a broken repository?

We converted a mercurial repository to git and things seemed to go well. We intially ran
```
git lfs migrate import --everything --include="*.dll,*.DLL,*.png"
```
on Linux, but all our users are on Windows. We noticed that there was an issue with some `*.DLL` files, so we redid the conversion to include both lowercase and uppercase library names, and put the repository online for weeks. It was only after quite a while when people started checking out ancient branches and commits that we noticed issues with some old `*.PNG` files that have since been deleted (so nobody noticed anything suspicious).

But now anyone trying to work with old branches, tags, etc. is facing the issue with
```
Encountered <N> file(s) that should have been pointers, but weren't:
```
We haven't yet decided whether or not to fix this (and force everyone to break their git clone) or live with the consequences and get used to use the `--force` flag when occasionally working with history.

But it would be awesome if `git lfs migrate` could at least throw a warning when it sees a file in git history that doesn't match the case-sensitive pattern (and doesn't get converted to large file), but would have matched on a case-insensitive file system.

**Comment by chrisd8088** [comment]
Hey, @mojca, thanks for the comment and I'm sorry you're having trouble.  With respect to your last idea:

> But it would be awesome if `git lfs migrate` could at least throw a warning when it sees a file in git history that doesn't match the case-sensitive pattern (and doesn't get converted to large file), but would have matched on a case-insensitive file system. 

This sounds like something it would be potentially useful to do, although one challenge would be to avoid spurious warnings for folks who are only going to be using case-sensitive file systems.  In any case, if you'd like to open a new issue to track this idea, that would be great, and that can then be a forum to discuss any implementation options.

**Comment by mojca** [comment]
@chrisd8088, thanks a lot for the feedback.

> although one challenge would be to avoid spurious warnings for folks who are only going to be using case-sensitive file systems.

Unless you are sure to remain the only user of a particular repository, it's basically impossible to predict when, say, a macOS user might end up using the repository.

Also, a "clean" repository (containing just `*.png` and no `*.PNG`) would usually not show any warnings. If all the files match the pattern exactly, no warning would ever show up. And even if the repository is only ever going to be used on a case sensitive system and the repo contains some `*.PNG` files, I suspect that many users requesting to turn `*.png` files into large files would prefer to see the former converted as well. So even if there would technically be no problem on that machine, the warning might be well in place.

The command is generally run once. I believe it's better to have false positives (warning that can be ignored) than false negatives (no warning for something the user might want to see).

> In any case, if you'd like to open a new issue to track this idea, that would be great, and that can then be a forum to discuss any implementation options.

Thank you. I opened a ticket (linked above).
